crates/bevy_ecs/src/system/commands/command_queue.rs

```diff
@@ -0,0 +1,181 @@
+use super::Command;
+use crate::world::World;
+
+/// # Safety
+///
+/// This function is only every called when the `command` bytes is the associated
+/// [`Commands`] `T` type. Also this only reads the data via `read_unaligned` so unaligned
+/// accesses are safe.
+unsafe fn invoke_command<T: Command>(command: *mut u8, world: &mut World) {
+    let command = if std::mem::size_of::<T>() > 0 {
+        command.cast::<T>().read_unaligned()
+    } else {
+        // NOTE: This is necessary because if the `CommandQueueInner` is only filled with 0-sized
+        // commands the `bytes` vec will never allocate. Then means that `bytes.as_ptr()` could be null
+        // and reading a null pointer is always UB.
+        // However according to https://doc.rust-lang.org/std/ptr/index.html
+        // "The canonical way to obtain a pointer that is valid for zero-sized accesses is NonNull::dangling"
+        // therefore the below code is safe to do.
+        let ptr = std::ptr::NonNull::<T>::dangling().as_ptr();
+        ptr.cast::<T>().read()
+    };
+    command.write(world);
+}
+
+struct CommandMeta {
+    offset: usize,
+    func: unsafe fn(value: *mut u8, world: &mut World),
+}
+
+#[derive(Default)]
+pub(crate) struct CommandQueueInner {
+    bytes: Vec<u8>,
+    metas: Vec<CommandMeta>,
+}
+
+// SAFE: All commands [`Command`] implement [`Send`]
+unsafe impl Send for CommandQueueInner {}
+
+// SAFE: `&CommandQueueInner` never gives access to the inner commands.
+unsafe impl Sync for CommandQueueInner {}
+
+impl CommandQueueInner {
+    /// Push a new `command` onto the queue.
+    #[inline]
+    pub fn push<C>(&mut self, command: C)
+    where
+        C: Command,
+    {
+        let size = std::mem::size_of::<C>();
```

Comment:
You might consider adding padding for alignment so that you can perform aligned reads above.

Reply:
I already went down that road actually haha :).
This current implementation ended up being far simpler to do, and I didn't really find any real benefit at the current time. 

Its especially difficult since if the byte vector reallocates you might need to re adjust for alignment. 

Imo, as it stands it is fine, however if there's a compelling argument to manually align everything (e.g. performance) we can probably explore that further.

Reply:
Also since references are never made to any of the commands, alignment is not needed luckily.

Reply:
> Its especially difficult since if the byte vector reallocates you might need to re adjust for alignment.

An arena or some kind of chunked allocator may perform better, regardless of the alignment/padding discussion.